pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -119,15 +117,22 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env model.Environmen
 			}
 		case model.ProtocolTCP, model.ProtocolMongo:
 			listenerType = plugin.ListenerTypeTCP
-			// TODO
-			// Look at virtual service specs, and identity destinations,
-			// call buildOutboundNetworkFilters.. and then construct TCPListener
-			//buildTCPListener(buildOutboundNetworkFilters(clusterName, addresses, servicePort),
-			//	listenAddress, uint32(servicePort.Port), servicePort.Protocol)
+			opts = buildListenerOpts{
+				env:        env,
+				proxy:      node,
+				ip:         WildcardAddress,
+				port:       int(server.Port.Number),
+				protocol:   model.ProtocolTCP,
+				sniHosts:   server.Hosts,
+				tlsContext: buildGatewayListenerTLSContext(server),
+				bindToPort: true,
+			}
+			networkFilters = buildGatewayInboundNetworkFilters(env, server)
```

Comment:
consider moving this inside oninboundlistener

Reply:
to what plugin though? Remember that we have to construct the basic set of filters per protocol. For HTTP, we add router and cors by ourselves. Similarly, for TCP, we need to add mongo and tcp. People can add their own in the callbacks

Reply:
call the plugin "base" maybe? the guts of building filters should be completely removed from gateway and listener imo.